### PR TITLE
Dra release staging stage

### PR DESCRIPTION
**Pull request by sharbuz**
## What is the problem this PR solves?

The partial migration of the Jenkins pipeline to the Buildkite pipeline. 

## How does this PR solve the problem?

Migration of the "DRA Release staging" step of the "packaging.groovy" pipeline.

## How to test this PR locally

It doesn't be checked locally, that's why it's a partial "step by step" migration

## Checklist

The Buildkite pipeline will be triggered by this PR and we will see the result

## Related issues

Related to the issue: https://github.com/elastic/fleet-server/issues/2517

**Pipeline steps:**
- [x] Pipeline init
- [x] Filter build - Skipped because this filter is required by Jenkins auto-scan and we already have all additional conditions for starting the pipeline
- [x] Package
- [x] DRA Snapshot
- [x] DRA Release Staging

### CODE REVIEW

File: .buildkite/pipeline.package.mbp.yml [hunk 1]
```diff
@@ -1,12 +1,56 @@
...
+        provider: "aws"
+        image: "core-ubuntu-2004-aarch64"
+      matrix:
+        setup:
+          type:
+          - "snapshot"
+          - "staging"
+
+  - label: "DRA snapshot"
+    key: "rda-snapshot"
```
Comment (mrodm) [inline .buildkite/pipeline.package.mbp.yml:37]: typo ? dra-snapshot ?
```suggestion
    key: "dra-snapshot"
```

File: .buildkite/pipeline.package.mbp.yml [hunk 2]
```diff
@@ -1,12 +1,56 @@
...
+    if: ${IS_FILE_EXISTS_IN_REPO}
+    depends_on:
+      - step: "package-publish"
+        allow_failure: false
 
-      - label: "step2 of the group"
-        command: echo "step2 of the group"
\ No newline at end of file
+  - label: "DRA release staging"
+    key: "rda-staging"
```
Comment (mrodm) [inline .buildkite/pipeline.package.mbp.yml:48]:
```suggestion
    key: "dra-staging"
```

File: .buildkite/hooks/pre-exit [hunk 3]
```diff
@@ -14,3 +14,11 @@ if [[ "$BUILDKITE_PIPELINE_SLUG" == "fleet-server" && "$BUILDKITE_STEP_KEY" == "
     unset GOOGLE_APPLICATIONS_CREDENTIALS
     cleanup
 fi
+
+if [[ "$BUILDKITE_PIPELINE_SLUG" == "fleet-server-package-mbp" ]]; then
+  if [[ "$BUILDKITE_STEP_KEY" == "package-x86-64" || "$BUILDKITE_STEP_KEY" == "package-arm" || "$BUILDKITE_STEP_KEY" == "rda-snapshot" && "$BUILDKITE_STEP_KEY" == "rda-staging" ]]; then
```
Comment (mrodm) [inline .buildkite/hooks/pre-exit:19]:
```suggestion
  if [[ "$BUILDKITE_STEP_KEY" == "package-x86-64" || "$BUILDKITE_STEP_KEY" == "package-arm" || "$BUILDKITE_STEP_KEY" == "dra-snapshot" && "$BUILDKITE_STEP_KEY" == "dra-staging" ]]; then
```

File: .buildkite/hooks/pre-command [hunk 4]
```diff
@@ -24,4 +27,22 @@ fi
...
+fi
+
+if [[ "$BUILDKITE_PIPELINE_SLUG" == "fleet-server-package-mbp" ]]; then
+  if [[ "$BUILDKITE_STEP_KEY" == "package-x86-64" || "$BUILDKITE_STEP_KEY" == "package-arm" || "$BUILDKITE_STEP_KEY" == "rda-snapshot" && "$BUILDKITE_STEP_KEY" == "rda-staging" ]]; then
+    export PRIVATE_CI_GCS_CREDENTIALS_SECRET=$(retry 5 vault kv get -field plaintext ${PRIVATE_CI_GCS_CREDENTIALS_PATH})
+    export JOB_GCS_BUCKET=$(retry 5 vault kv get -field plaintext ${JOB_GCS_BUCKET_SECRET_PATH})
+  fi
+fi
+
+if [[ "$BUILDKITE_PIPELINE_SLUG" == "fleet-server-package-mbp" && "$BUILDKITE_STEP_KEY" == "rda-snapshot" && "$BUILDKITE_STEP_KEY" == "rda-staging" ]]; then
```
Comment (mrodm) [inline .buildkite/hooks/pre-command:40]:
```suggestion
if [[ "$BUILDKITE_PIPELINE_SLUG" == "fleet-server-package-mbp" && "$BUILDKITE_STEP_KEY" == "dra-snapshot" && "$BUILDKITE_STEP_KEY" == "dra-staging" ]]; then
```

File: .buildkite/hooks/post-checkout [hunk 5]
```diff
@@ -32,6 +34,49 @@ checkout_merge() {
...
+    if [[ $branch == "main" ]]; then
+        local branchName="master"
+    fi
+
+    local path_to_file="cd/release/release-manager/project-configs/${branchName}/${fileName}"
+    local response=$(curl -s -u $GITHUB_REPO_TOKEN:x-oauth-basic https://api.github.com/repos/elastic/${repoName}/contents/${path_to_file}| grep -c "\"path\"\: \"${path_to_file}\"")
+    if [[ ${response} -ge 1 ]]; then
+        export IS_FILE_EXISTS_IN_REPO=true
+    else
+        export IS_FILE_EXISTS_IN_REPO=false
```
Comment (mrodm) [inline .buildkite/hooks/post-checkout:76]: This variable is the same as this one ? 
https://github.com/elastic/fleet-server/blob/331112b16439ccfcef7883e49baf5ac325515601/.ci/packaging.groovy#L122

Why not calling it as in Jenkins `IS_BRANCH_AVAILABLE`?

File: .buildkite/pipeline.package.mbp.yml [hunk 6]
```diff
@@ -1,12 +1,56 @@
...
+          - "snapshot"
+          - "staging"
+
+  - label: "DRA snapshot"
+    key: "rda-snapshot"
+    command: ".buildkite/scripts/dra_release.sh snapshot"
+    agents:
+      provider: "gcp"
+      image: "family/core-ubuntu-2004"
+    if: ${IS_FILE_EXISTS_IN_REPO}
```
Comment (mrodm) [inline .buildkite/pipeline.package.mbp.yml:42]: Not sure this variable is available at this moment.

This variable is calculated in the post-checkout hook, and this hook is executed once the step is already running.

Would it be neede to move those functions that calculate this variable into the `.buildkite/scripts/dra_release.sh` script ?

File: .buildkite/pipeline.package.mbp.yml [hunk 7]
```diff
@@ -1,12 +1,56 @@
...
\ No newline at end of file
+  - label: "DRA release staging"
+    key: "rda-staging"
+    command: ".buildkite/scripts/dra_release.sh staging"
+    agents:
+      provider: "gcp"
+      image: "family/core-ubuntu-2004"
+    if: "${IS_FILE_EXISTS_IN_REPO} == true && build.env('BUILDKITE_BRANCH') != 'main'"
+    depends_on:
+      - step: "rda-snapshot"
```
Comment (mrodm) [inline .buildkite/pipeline.package.mbp.yml:55]:
```suggestion
      - step: "dra-snapshot"
```

File: .buildkite/pipeline.package.mbp.yml [hunk 8]
```diff
@@ -1,12 +1,56 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/buildkite/pipeline-schema/main/schema.json
 name: "fleet server package mbp"
```
Comment (mrodm) [inline .buildkite/pipeline.package.mbp.yml:2]: This name is overwriting the name or description in the catalog-info.yaml file ?

File: .buildkite/scripts/common.sh [hunk 9]
```diff
@@ -6,15 +6,21 @@ WORKSPACE="$(pwd)/bin"
 TMP_FOLDER_TEMPLATE_BASE="tmp.fleet-server"
 REPO="fleet-server"
 
-add_bin_path(){
-    echo "Adding PATH to the environment variables..."
+create_workspace() {
+    if [[ ! -d "${WORKSPACE}" ]]; then
     mkdir -p ${WORKSPACE}
+    fi
```
Comment (mrodm) [inline .buildkite/scripts/common.sh:12]: Maybe it can be removed the if condition.
`mkdir -p` does not fail if `WORKSPACE` folder exists. But it could be kept if you prefer.

If kept the `if` block, just a quick suggestion
```suggestion
    if [[ ! -d "${WORKSPACE}" ]]; then
        mkdir -p ${WORKSPACE}
    fi
```

File: .buildkite/scripts/common.sh [hunk 10]
```diff
@@ -94,9 +100,50 @@ upload_packages_to_gcp_bucket() {
...
+            "github.com/magefile/mage"
+            "github.com/elastic/go-licenser"
+            "golang.org/x/tools/cmd/goimports"
+            "github.com/jstemmer/go-junit-report"
+            "gotest.tools/gotestsum"
+    )
+    create_workspace
+    for pkg in "${install_packages[@]}"; do
+    go install "${pkg}@latest"
+    done
```
Comment (mrodm) [inline .buildkite/scripts/common.sh:140]: nit
```suggestion
    for pkg in "${install_packages[@]}"; do
        go install "${pkg}@latest"
    done
```

File: .buildkite/scripts/common.sh [hunk 11]
```diff
@@ -94,9 +100,50 @@ upload_packages_to_gcp_bucket() {
...
+}
+
 cleanup() {
-    echo "Deleting temporal files..."
-    cd ${WORKSPACE}
-    rm -rf ${TMP_FOLDER_TEMPLATE_BASE}.*
+    echo "Deleting temporary files..."
+    if [[ ! -d "${WORKSPACE}/${TMP_FOLDER_TEMPLATE_BASE}.*" ]]; then
+        rm -rf ${WORKSPACE}/${TMP_FOLDER_TEMPLATE_BASE}.*
+    fi
```
Comment (mrodm) [inline .buildkite/scripts/common.sh:147]: I've tried these commands, and it loks like `-d` with a glob does not work.
```
 $ mkdir -p test.12345
 
 $ if [[ -d "test.*" ]] ; then echo "found"; else echo "not found"; fi
not found
```

As `rm -rf`  would not fail in case that template folder (with glob pattern) does not exist, I would suggest to just keep the `rm -rf` without the `if`

```suggestion
    rm -rf ${WORKSPACE}/${TMP_FOLDER_TEMPLATE_BASE}.*
```

File: .buildkite/scripts/package.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    with_mage
+
+    if [[ ${MATRIX_TYPE} == "staging" ]]; then
+        make release
+    else
+        make SNAPSHOT=true release
+    fi
+    google_cloud_auth
+    upload_mbp_packages_to_gcp_bucket "build/distributions/" "${MATRIX_TYPE}"
+fi
```
Comment (mrodm) [inline .buildkite/scripts/package.sh:34]: No need of the `else` block since if it is true, it already runs `exit 0`

```suggestion
if [[ ${BUILDKITE_BRANCH} == "main" && ${MATRIX_TYPE} == "staging" ]]; then
    echo "INFO: staging artifacts for the main branch are not required."
    exit 0
PLATFORMS=""
PACKAGES=""
if [[ ${PLATFORM_TYPE} == "arm" || ${PLATFORM_TYPE} == "aarch64" ]]; then
    PLATFORMS="linux/arm64"
    PACKAGES="docker"
fi

add_bin_path
with_go
with_mage

if [[ ${MATRIX_TYPE} == "staging" ]]; then
    make release
else
    make SNAPSHOT=true release
fi

google_cloud_auth
upload_mbp_packages_to_gcp_bucket "build/distributions/" "${MATRIX_TYPE}"
```

File: .buildkite/scripts/dra_release.sh [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+if [[ "${VERSION}" == *"-SNAPSHOT"* || "${VERSION}" == "" ]]; then
+    echo "The 'version' parameter is required and it cannot contain the suffix '-SNAPSHOT'."
+    exit 1
+fi
+
+if [[ "${PROJECT}" == "" ]]; then
+    echo "The 'project' parameter is required."
+    exit 1
+fi
```
Comment (mrodm) [inline .buildkite/scripts/dra_release.sh:23]: PROJECT is already set as variable in line 11 in this script. Should it be moved as a env as part of the step definition in the pipeline yaml?

Why is it needed this safeguard? Is this script intended to be copied into other repositories ?

File: .buildkite/hooks/post-checkout [hunk 14]
```diff
@@ -32,6 +34,49 @@ checkout_merge() {
...
+            sleep $wait
+        else
+            >&2 echo "Retry $count/$retries exited $exit, no more retries left."
+            return $exit
+        fi
+    done
+    return 0
+}
+
+GITHUB_REPO_TOKEN=$(retry 5 vault kv get -field token ${GITHUB_TOKEN_VAULT_PATH})
```
Comment (mrodm) [inline .buildkite/hooks/post-checkout:57]: Usually secrets are retrieved in the `pre-command` hook. Why are those needed here?

File: .buildkite/hooks/post-checkout [hunk 15]
```diff
@@ -32,6 +34,49 @@ checkout_merge() {
...
+    local repoName=${1}
+    local branch=${2}
+    local fileName="build.gradle"
+
+    if [[ $branch == "main" ]]; then
+        local branchName="master"
+    fi
+
+    local path_to_file="cd/release/release-manager/project-configs/${branchName}/${fileName}"
+    local response=$(curl -s -u $GITHUB_REPO_TOKEN:x-oauth-basic https://api.github.com/repos/elastic/${repoName}/contents/${path_to_file}| grep -c "\"path\"\: \"${path_to_file}\"")
```
Comment (mrodm) [inline .buildkite/hooks/post-checkout:72]: Does this function requires to have a working copy of `fleet-server` in-place ? 

If don't, I would consider to move the Github secrets and this function into the pre-command hook.

File: .buildkite/scripts/release-manager.sh [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+# It uses env variables to help to run this script with a simpler jenkins
+# pipeline call.
+#
+set -uexo pipefail
+
+readonly TYPE=${TYPE:-snapshot}
+readonly OUTPUT_FILE=${OUTPUT_FILE:-release-manager-report.out}
+
+# shellcheck source=/dev/null
+source /usr/local/bin/bash_standard_lib.sh
```
Comment (mrodm) [inline .buildkite/scripts/release-manager.sh:24]: Does this file exists on buildkite agents ?

File: .buildkite/scripts/dra_release.sh [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+if [[ "${TYPE}" == "snapshot" ]]; then
+    SNAPSHOT=true
+fi
+
+mkdir -p ${BASE_DIR}/reports
+./dev-tools/dependencies-report --csv ${BASE_DIR}/reports/dependencies-${VERSION}.csv
+cd ${BASE_DIR}/reports && shasum -a 512 dependencies-${VERSION}.csv > dependencies-${VERSION}.csv.sha512
+
+echo "test run before running the ./scripts/release-manager.sh script" #TODO - chould be chaged to "./scripts/release-manager.sh" after the aprooval
```
Comment (mrodm) [inline .buildkite/scripts/dra_release.sh:43]: Was available a `dry-run` mode in this script ? cc @alexsapran 
Shall we just keep for the time being commented out this script ?

File: .buildkite/scripts/dra_release.sh (hunk 13 above)
Comment (sharbuz) [reply .buildkite/scripts/dra_release.sh:23]: It was implemented in the original pipeline with hard-coded values. And It looks like a double-checking because it's an important stage.

File: .buildkite/hooks/post-checkout (hunk 14 above)
Comment (sharbuz) [reply .buildkite/hooks/post-checkout:57]: The answers here: 
https://github.com/elastic/fleet-server/pull/2702#discussion_r1239715942 
and here:
https://github.com/elastic/fleet-server/pull/2702#discussion_r1239721118

File: .buildkite/hooks/post-checkout (hunk 15 above)
Comment (sharbuz) [reply .buildkite/hooks/post-checkout:72]: https://github.com/elastic/fleet-server/pull/2723#discussion_r1239748606

File: .buildkite/scripts/dra_release.sh (hunk 17 above)
Comment (sharbuz) [reply .buildkite/scripts/dra_release.sh:43]: Nope, the dry-run mode wasn't available here before and, if I'm not wrong we decided to add this mode at least for tests till the final migration.

File: .buildkite/hooks/post-checkout (hunk 5 above)
Comment (sharbuz) [reply .buildkite/hooks/post-checkout:76]: It was renamed because, in fact, it's checking a file in the repository, not a branch.

File: .buildkite/hooks/post-checkout [hunk 18]
```diff
@@ -6,6 +6,8 @@ source .buildkite/scripts/common.sh
 
 GITHUB_TOKEN_VAULT_PATH="kv/ci-shared/platform-ingest/github_token"
 GITHUB_REPO_TOKEN=$(retry 5 vault kv get -field token ${GITHUB_TOKEN_VAULT_PATH})
+#export GO_VERSION=$(retry 5 curl -s https://api.github.com/repos/elastic/fleet-server/contents/.go-version | jq -r ".content" | base64 --decode)
+export GO_VERSION=$(retry 5 curl -s -H 'Accept: application/vnd.github.v3.raw' https://api.github.com/repos/elastic/fleet-server/contents/.go-version)
```
Comment (mrodm) [inline .buildkite/hooks/post-checkout:10]: Which reference is used here to retrieve the contents of the file ? main ?

It needs to be used the contents of the actual branch. Even, PR branches could just exist in other remotes.

File: .buildkite/scripts/common.sh [hunk 19]
```diff
@@ -69,35 +68,45 @@ docker_logout() {
...
+    local baseUri="gs://${JOB_GCS_BUCKET}/${REPO}/buildkite"              #TODO: needs to delete the "/buildkite" part after the migration from Jenkins
+    local bucketUriCommit="${baseUri}"/commits/${BUILDKITE_COMMIT}
+    local bucketUriDefault="${baseUri}"/snapshots
 
     if [[ ${BUILDKITE_PULL_REQUEST} != "false" ]]; then
         bucketUriDefault="${baseUri}"/pull-requests/pr-${GITHUB_PR_NUMBER}
     fi
+        for bucketUri in "${bucketUriCommit}" "${bucketUriDefault}"; do
+            gsutil -m -q cp -a public-read -r ${pattern} "${bucketUri}"
+        done
```
Comment (mrodm) [inline .buildkite/scripts/common.sh:98]: nit. Just indentation.
```suggestion
    for bucketUri in "${bucketUriCommit}" "${bucketUriDefault}"; do
        gsutil -m -q cp -a public-read -r ${pattern} "${bucketUri}"
    done
```

File: .buildkite/scripts/release-manager.sh [hunk 20]
```diff
@@ -20,9 +20,6 @@ set -uexo pipefail
 readonly TYPE=${TYPE:-snapshot}
 readonly OUTPUT_FILE=${OUTPUT_FILE:-release-manager-report.out}
 
-# shellcheck source=/dev/null
-source /usr/local/bin/bash_standard_lib.sh
```
Comment (mrodm) [inline .buildkite/scripts/release-manager.sh:24]: What was the needed function from this file ?

Comment (sharbuz) [reply .buildkite/scripts/release-manager.sh:24]: Why are seeing that?
It was removed days ago...

File: .buildkite/hooks/post-checkout (hunk 18 above)
Comment (sharbuz) [reply .buildkite/hooks/post-checkout]: resolved with:
`export GO_VERSION=$(cat .go-version)`

File: .buildkite/pipeline.yml [hunk 21]
```diff
@@ -3,7 +3,6 @@
 env:
   SETUP_GVM_VERSION: 'v0.5.0'
   DOCKER_COMPOSE_VERSION: '1.25.5'
-  GO_VERSION: "1.20.5"
```
[the remark this answers is not in the document]
Comment (sharbuz) [reply .buildkite/pipeline.yml:6]: to be honest, I wasn't involved in this part of the code and it requires to spend some time to understand this part and why I should change something in the `.ci` folder, I supposed my work directory is `.buildkite`...

[the remark this answers is not in the document]
Comment (sharbuz) [reply .buildkite/pipeline.yml:6]: Hi @mrodm, got it, thanks for clarifying.
done, please check it once again.

[the remark this answers is not in the document]
Comment (mrodm) [reply .buildkite/pipeline.yml:6]: Thanks @sharbuz

**Review by mrodm — APPROVED** [review]
:+1:
